FirebaseMessaging/Tests/UnitTests/FIRMessagingAuthServiceTest.m

```diff
@@ -0,0 +1,400 @@
+  XCTestExpectation *checkinExpectation =
+      [self expectationWithDescription:@"Did call checkin service"];
+
+  FIRMessagingCheckinPreferences *checkinPreferences = [self validCheckinPreferences];
+  [[[self.mockCheckinService stub] andDo:^(NSInvocation *invocation) {
+    self.checkinCompletion(checkinPreferences, nil);
+  }] checkinWithExistingCheckin:[OCMArg any]
+                     completion:[OCMArg checkWithBlock:^BOOL(id obj) {
+                       [checkinExpectation fulfill];
+                       self.checkinCompletion = obj;
+                       return obj != nil;
+                     }]];
+
+  // Always return YES for whether we succeeded in persisting the checkin
+  [[self.mockStore stub] saveCheckinPreferences:[OCMArg any]
+                                        handler:[OCMArg invokeBlockWithArgs:[NSNull null], nil]];
+
+  [self.authService scheduleCheckin:YES];
+
+  XCTAssertTrue([self.authService hasValidCheckinInfo]);
+  XCTAssertEqual([self.authService checkinRetryCount], 1);
+  [self waitForExpectationsWithTimeout:2.0 handler:NULL];
+}
+
+/**
+ *  Test scheduling a checkin which completes successfully, but fails to save, due to Keychain
+ *  errors.
+ */
+- (void)testScheduleCheckin_successButFailureInSaving {
+  XCTestExpectation *checkinFailureExpectation =
+      [self expectationWithDescription:@"Did receive error after checkin"];
+
+  FIRMessagingCheckinPreferences *checkinPreferences = [self validCheckinPreferences];
+  [[[self.mockCheckinService stub] andDo:^(NSInvocation *invocation) {
+    self.checkinCompletion(checkinPreferences, nil);
+  }] checkinWithExistingCheckin:[OCMArg any]
+                     completion:[OCMArg checkWithBlock:^BOOL(id obj) {
+                       self.checkinCompletion = obj;
+                       return obj != nil;
+                     }]];
+
+  // Always return NO for whether we succeeded in persisting the checkin, to simulate Keychain error
+  [[self.mockStore stub] saveCheckinPreferences:[OCMArg any]
+                                        handler:[OCMArg invokeBlockWithArgs:[OCMArg any], nil]];
+
+  [self.authService
+      fetchCheckinInfoWithHandler:^(FIRMessagingCheckinPreferences *checkin, NSError *error) {
+        [checkinFailureExpectation fulfill];
+      }];
+
+  [self waitForExpectationsWithTimeout:2.0 handler:NULL];
+  XCTAssertFalse([self.authService hasValidCheckinInfo]);
+}
+
+/**
+ *  Test scheduling multiple checkins to complete immediately. Each successive checkin should
+ *  be triggered immediately.
+ */
+- (void)testMultipleScheduleCheckin_immediately {
+  XCTestExpectation *checkinExpectation =
+      [self expectationWithDescription:@"Did call checkin service"];
+  __block int checkinHandlerInvocationCount = 0;
+
+  FIRMessagingCheckinPreferences *checkinPreferences = [self validCheckinPreferences];
+  [[[self.mockCheckinService stub] andDo:^(NSInvocation *invocation) {
+    checkinHandlerInvocationCount++;
```

Comment:
It looks redundant as `checkinExpectation` should assert it when [assertForOverFulfill == YES](https://developer.apple.com/documentation/xctest/xctestexpectation/2806575-assertforoverfulfill?language=objc).

Reply:
I'm not following this one, the existing test is not using assertForOverFulfill and checkinExpectation seems required here.

Reply:
We can add todo if you insist this should be fixed, but these are really old tests and we try to first move all of them over first to ensure all token logics/behaviors are intact with original IID tests.

Reply:
TL;DR Actually all LGTM here.

assertForOverFulfill is YES by default, so my initial thought was that if the method is called more than once the expectation will eventually be over fulfilled and the test would fail. But it looks like due to the fulfilment delay we still need to keep track of `checkinHandlerInvocationCount` to be 100% sure.